source/common/conn_pool/conn_pool_base.cc

```diff
@@ -432,7 +474,9 @@ void ConnPoolImplBase::onConnectionEvent(ActiveClient& client, absl::string_view
       client.connect_timer_->disableTimer();
       client.connect_timer_.reset();
     }
+    const bool contributed_to_connecting_stream_capacity = client.currentUnusedCapacity() > 0;
```

Comment:
I'm sure this is correct and probably obvious but I can't quite seem to understand why unused capacity being greater than 0 means it contributed to stream capacity. Can you shed light? Maybe I don't quite understand when the capacity changes.

Reply:
sorry you're saying this handles the case where an early data stream had capacity 10, had 10 streams assigned, then disconnected?
I agree but I don't think this is the way to handle it.  I think when we have a failure to connect, we should purge N streams, where N was the capacity of the one which failed to connect. I'd be inclined to take this out and in a follow up Pr have purge pending streams only purge connecting capacity worth.

if you've got 6 pending streams and 6 HTTP/1.1 connections there's no point killing all 6 streams when one HTTP connection fails.

Reply:
Are you suggesting we change this to only purge the remaining connecting capacity of the failed client? I.e. a TLS early data client with capacity 10 with 1 early data stream. If its handshake times out, it should purge upto 9 pending streams if there is any?

Reading the comment below, it seems like purging all pending stream is intentional though:
```// Raw connect failures should never happen under normal circumstances. If we have an
      // upstream that is behaving badly, streams can get stuck here in the pending state. If we
      // see a connect failure, we purge all pending streams so that calling code can determine
      // what to do with the stream.```

Reply:
I think the problem was if you had one connection and pending streams associated, and that connection failed, the streams would be in limbo forever.
I think if you have 2 pending streams (say for HTTP/1) and 2 pending connections, when you lose a connection you should lose only one stream.  Fixing that is outside of the scope of this PR but I don't think we want to add more complexity here when the problem extends beyond HTTP/3.